make the tx working again

mikedoan opened the pull request:
make the tx working again

LeoHChen commented on node/worker/worker.go:
```diff
@@ -67,6 +67,9 @@ func (w *Worker) SelectTransactionsForNewBlock(txs types.Transactions, maxNumTxs
 		}
 	}
+	if err := w.UpdateCurrent(coinbase); err != nil {
+		utils.GetLogger().Debug("Failed updating worker's state", "Error", err)
```
Is it debug or error?

LeoHChen commented on node/node_newblock.go:
```diff
@@ -80,11 +80,6 @@ func (node *Node) WaitForConsensusReadyv2(readySignal chan struct{}, stopChan ch
 					viewID := node.Consensus.GetViewID()
 					// add aggregated commit signatures from last block, except for the first two blocks
 
-					err = node.Worker.UpdateCurrent(coinbase)
```
who made this change?  what's the reason of making this change here?

mikedoan replied:
RJ made. i synced with him today.

mikedoan replied:
Somehow if i remove the commitTransactions - tx is not working anymore

LeoHChen commented on core/types/transaction.go:
```diff
@@ -50,8 +50,7 @@ type txdata struct {
 	Price        *big.Int        `json:"gasPrice" gencodec:"required"`
 	GasLimit     uint64          `json:"gas"      gencodec:"required"`
 	ShardID      uint32          `json:"shardID"  gencodec:"required"`
-	ToShardID    uint32          `json:"toShardID" rlp:"nil"` // for cross-shard tx's destination shard ID; nil means intra-shard tx
```
why remove ToShardID? we have discussed in discord.

mikedoan replied:
this is PR for john. if i remove that and use the new version of the sdk, it's not working. i created this for John.

LeoHChen commented:
The original change was to fix the block number mismatch issue.
```
commit d7aa1d1adf0ae15971f55ba29780071e4c550d4c
Author: Rongjian Lan <rongjian.lan@gmail.com>
Date:   Sat Jun 22 22:25:13 2019 -0700

    Fix block num mismatch issue
```

mikedoan commented:
> The original change was to fix the block number mismatch issue.
> 
> ```
> commit d7aa1d1adf0ae15971f55ba29780071e4c550d4c
> Author: Rongjian Lan <rongjian.lan@gmail.com>
> Date:   Sat Jun 22 22:25:13 2019 -0700
> 
>     Fix block num mismatch issue
> ```

yes - that was incorrect change.

actually this new is actually not correct one too. it's a long investigation.

mikedoan commented on node/worker/worker.go:
```diff
@@ -67,6 +67,9 @@ func (w *Worker) SelectTransactionsForNewBlock(txs types.Transactions, maxNumTxs
 			}
 		}
 	}
+	if err := w.UpdateCurrent(coinbase); err != nil {
```
In the past, before disabling the tx for mainnet - the logic here is to call getTransactionsForNewBlock first https://github.com/harmony-one/harmony/blob/before_sc/node/node_newblock.go#L140, then call Worker.CommitTransactions

And inside getTransactionsForNewBlock, we call w.UpdateCurrent()

During refactoring for mainnet - we removed w.UpdateCurrent() out of getTransactionsForNewBlock and moved it after CommitTransactions.

I did a lot of debug and found out we need to move it back.

I also found CommitTransactions redundant to our code - and tried to removed it but removing that made the tx logic not working again.

Since I had to progress on SDK, integration test, SDK workflow i did continue the investigation to answer why removing CommitTransactions made tx logic not working.

mikedoan commented:
In the past, before disabling the tx for mainnet - the logic here is to call getTransactionsForNewBlock first https://github.com/harmony-one/harmony/blob/before_sc/node/node_newblock.go#L140, then call Worker.CommitTransactions. And inside getTransactionsForNewBlock, we call w.UpdateCurrent() <-- [1]

During refactoring for mainnet - we removed w.UpdateCurrent() out of getTransactionsForNewBlock and moved it after CommitTransactions. <-- [2]

I did a lot of debug and found out we need to revert [1] back.

mikedoan commented:
Jenkins, test this please.

mikedoan commented:
Jenkins, test this please.

mikedoan commented:
Jenkins, test this please.

rlan35 commented on cmd/harmony/main.go:
```diff
@@ -264,7 +264,7 @@ func createGlobalConfig(isLeader bool) *nodeconfig.ConfigType {
 	}
 
 	nodeConfig.Host, err = p2pimpl.NewHost(&nodeConfig.SelfPeer, nodeConfig.P2pPriKey)
-	if *logConn && nodeConfig.GetNetworkType() != nodeconfig.Mainnet {
```
We don’t wanna log for mainnet, it will blow up the log size.

rlan35 commented on cmd/harmony/main.go:
```diff
@@ -480,11 +480,9 @@ func main() {
 	currentNode.ServiceManagerSetup()
 
-	// RPC for SDK not supported for mainnet.
-	if *networkType != nodeconfig.Mainnet {
```
Why remove this condition?

mikedoan replied:
because from now we want to enable the sdk by default

LeoHChen commented on core/types/transaction.go:
```diff
@@ -50,8 +50,7 @@ type txdata struct {
 	Price        *big.Int        `json:"gasPrice" gencodec:"required"`
 	GasLimit     uint64          `json:"gas"      gencodec:"required"`
 	ShardID      uint32          `json:"shardID"  gencodec:"required"`
-	ToShardID    uint32          `json:"toShardID" rlp:"nil"` // for cross-shard tx's destination shard ID; nil means intra-shard tx
```
why remove ToShardID?  we need it for cross-shard tx.

mikedoan replied:
i talked with rj, and synced with richard and michael. we will enable it back when implementing it. i know after the dinner meeting today, we would probably implement the cross shard soon.
